src/env.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An environment which is passed downward into the widget tree.
+
+use std::collections::HashMap;
+use std::fmt::{Debug, Formatter};
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use crate::kurbo::{Point, Rect, Size};
+use crate::piet::Color;
+
+use crate::Data;
+
+/// An environment passed down through all widget traversals.
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone, Default)]
+pub struct Env(Arc<EnvImpl>);
+
+#[derive(Clone, Default)]
+struct EnvImpl {
+    map: HashMap<String, Value>,
+}
+
+/// A typed key.
+///
+/// This lets you retrieve values of a given type. The parameter
+/// implements [`ValueType`]. For "expensive" types, this is a reference,
+/// so the type for a string is `Key<&str>`.
+///
+/// [`ValueType`]: trait.ValueType.html
+pub struct Key<T> {
+    key: &'static str,
+    value_type: PhantomData<T>,
+}
+
+// we could do some serious deriving here: the set of types that can be stored
+// could be defined per-app
+// Also consider Box<Any> (though this would also impact debug).
+/// A dynamic type representing all values that can be stored in an environment.
+#[derive(Clone)]
+pub enum Value {
+    Point(Point),
+    Size(Size),
+    Rect(Rect),
+    Color(Color),
+    Float(f64),
+    String(String),
+}
+
+/// Values which can be stored in an environment.
+///
+/// Note that for "expensive" types this is the reference. For example,
+/// for strings, this trait is implemented on `&'a str`. The trait is
+/// parametrized on a lifetime so that it can be used for references in
+/// this way.
+pub trait ValueType<'a>: Sized {
+    /// The corresponding owned type.
+    type Owned;
+
+    /// Attempt to convert the generic `Value` into this type.
+    fn try_from_value(v: &'a Value) -> Result<Self, String>;
```

Comment:
We'll need to figure out an error type for this at some point; for now I don't mind this being a `TODO`.

Reply:
I'll typedef it with a TODO.